druid-shell/src/platform/x11/window.rs

```diff
@@ -299,6 +398,58 @@ atom_manager! {
 /// The mutable state of the window.
 struct WindowState {
     size: Size,
+    /// The region that was invalidated since the last time we rendered.
+    invalid: Rect,
+}
+
+/// A collection of pixmaps for rendering to. This gets used in two different ways: if the present
+/// extension is enabled, we render to a pixmap and then present it. If the present extension is
+/// disabled, we render to a pixmap and then call `copy_area` on it (this probably isn't the best
+/// way to imitate double buffering, but it's the fallback anyway).
+struct Buffers {
+    /// A list of idle pixmaps. We take a pixmap from here for rendering to.
+    ///
+    /// When we're not using the present extension, all pixmaps belong in here; as soon as we copy
+    /// from one, we can use it again.
+    ///
+    /// When we submit a pixmap to present, we're not allowed to touch it again until we get a
+    /// corresponding IDLE_NOTIFY event. In my limited experiments this happens shortly after
+    /// vsync, meaning that we may want to start rendering the next pixmap before we get the old
+    /// one back. Therefore, we keep a list of pixmaps. We pop one each time we render, and push
+    /// one when we get IDLE_NOTIFY.
+    ///
+    /// Since the current code only renders at most once per vsync, two pixmaps seems to always be
+    /// enough. Nevertheless, we will allocate more on the fly if we need them. Note that rendering
+    /// more than once per vsync can only improve latency, because only the most recently-presented
+    /// pixmap will get rendered.
+    idle_pixmaps: Vec<Pixmap>,
+    /// A list of all the allocated pixmaps (including the idle ones).
+    all_pixmaps: Vec<Pixmap>,
+    /// The sizes of the pixmaps (they all have the same size). In order to avoid repeatedly
+    /// reallocating as the window size changes, we allow these to be bigger than the window.
+    width: u16,
+    height: u16,
+    /// The depth of the currently allocated pixmaps.
+    depth: u8,
+}
+
+/// The state involved in using X's Present extension.
+#[derive(Debug)]
+struct PresentData {
+    /// A monotonically increasing request counter.
+    serial: u32,
```

Comment:
Just to make it slightly more explicit:
```suggestion
    /// A monotonically increasing present request counter.
    serial: u32,
```